File: packages/oxlint-plugin-react-doctor/src/plugin/rules/design/no-inert-pointer-affordance.ts
import { HTML_TAGS } from "../../constants/html-tags.js";
import { defineRule } from "../../utils/define-rule.js";
import { getUnvariantClassNameTokens } from "../../utils/get-unvariant-class-name-tokens.js";
import type { EsTreeNode } from "../../utils/es-tree-node.js";
import type { EsTreeNodeOfType } from "../../utils/es-tree-node-of-type.js";
import { getJsxAttributeName } from "../../utils/get-jsx-attribute-name.js";
import { getStringLiteralAttributeValue } from "../../utils/get-string-literal-attribute-value.js";
import { hasJsxPropIgnoreCase } from "../../utils/has-jsx-prop-ignore-case.js";
import { hasJsxSpreadAttribute } from "../../utils/has-jsx-spread-attribute.js";
import { isInteractiveElement } from "../../utils/is-interactive-element.js";
import { isInteractiveRole } from "../../utils/is-interactive-role.js";
import { isNodeOfType } from "../../utils/is-node-of-type.js";
import { resolveJsxElementType } from "../../utils/resolve-jsx-element-type.js";
import type { RuleContext } from "../../utils/rule-context.js";
import { getStringFromClassNameAttr } from "./utils/get-string-from-class-name-attr.js";
import { getLastMatchingToken } from "./utils/get-last-matching-token.js";

const POINTER_HANDLER_PATTERN = /^on(?:click|pointer|mouse|touch|drag)/i;

const hasPointerBehaviorSignal = (
  openingElement: EsTreeNodeOfType<"JSXOpeningElement">,
): boolean => {
  if (hasJsxSpreadAttribute(openingElement.attributes)) return true;
  if (
    hasJsxPropIgnoreCase(openingElement.attributes, "tabIndex") ||
    hasJsxPropIgnoreCase(openingElement.attributes, "draggable") ||
    hasJsxPropIgnoreCase(openingElement.attributes, "contentEditable")
  ) {
    return true;
  }
  for (const attribute of openingElement.attributes) {
    if (!isNodeOfType(attribute, "JSXAttribute")) continue;
    const attributeName = getJsxAttributeName(attribute.name);
    if (attributeName && POINTER_HANDLER_PATTERN.test(attributeName)) return true;
  }
  const roleAttribute = hasJsxPropIgnoreCase(openingElement.attributes, "role");
  if (!roleAttribute) return false;
  const roleValue = getStringLiteralAttributeValue(roleAttribute);
  if (roleValue === null) return true;
  const firstRole = roleValue.trim().toLowerCase().split(/\s+/)[0];
  return Boolean(firstRole && isInteractiveRole(firstRole));
};

const isInteractionBoundary = (openingElement: EsTreeNodeOfType<"JSXOpeningElement">): boolean => {
  const elementType = resolveJsxElementType(openingElement);
  return (
    elementType === "label" ||
    (HTML_TAGS.has(elementType) && isInteractiveElement(elementType, openingElement)) ||
    hasPointerBehaviorSignal(openingElement)
  );
};

const hasWrappingInteractionBoundary = (
  openingElement: EsTreeNodeOfType<"JSXOpeningElement">,
): boolean => {
  let ancestor: EsTreeNode | null | undefined = openingElement.parent?.parent;
  while (ancestor) {
    if (isNodeOfType(ancestor, "JSXAttribute")) {
      const isChildrenAttribute =
        isNodeOfType(ancestor.name, "JSXIdentifier") && ancestor.name.name === "children";
      if (!isChildrenAttribute) return false;
    }
    if (isNodeOfType(ancestor, "JSXElement") && isInteractionBoundary(ancestor.openingElement)) {
      return true;
    }
    ancestor = ancestor.parent;
  }
  return false;
};

export const noInertPointerAffordance = defineRule({
  id: "no-inert-pointer-affordance",
  title: "Pointer cursor has no interaction",
  severity: "warn",
  defaultEnabled: false,
  tags: ["design", "test-noise", "react-jsx-only"],
  requires: ["tailwind"],
  category: "Accessibility",
  recommendation:
    "Remove the pointer cursor or put the affordance on the control that actually handles the interaction.",
  create: (context: RuleContext) => ({
    JSXOpeningElement(node: EsTreeNodeOfType<"JSXOpeningElement">) {
      const elementType = resolveJsxElementType(node);
      if (!HTML_TAGS.has(elementType) || isInteractiveElement(elementType, node)) return;
      const classNameValue = getStringFromClassNameAttr(node);
      if (!classNameValue) return;
      const cursorUtility = getLastMatchingToken(
        getUnvariantClassNameTokens(classNameValue),
        (utility) => utility.startsWith("cursor-"),
      );
      if (cursorUtility !== "cursor-pointer") return;
      if (elementType === "label" || hasPointerBehaviorSignal(node)) return;
      if (hasWrappingInteractionBoundary(node)) return;
      context.report({
        node,
        message:
          "This pointer cursor promises an interaction, but neither this element nor its wrapping surface handles one.",
      });
    },
  }),
});

File: native/oxlint/rules/no-inert-pointer-affordance.rs
use oxc_ast::{AstKind, ast::JSXAttributeItem};
use oxc_diagnostics::OxcDiagnostic;
use oxc_macros::declare_oxc_lint;

use crate::{
    AstNode,
    context::{ContextHost, LintContext},
    globals::HTML_TAG,
    rule::Rule,
    utils::{has_jsx_prop_ignore_case, is_interactive_element, is_interactive_role},
};

const MESSAGE: &str = "This pointer cursor promises an interaction, but neither this element nor its wrapping surface handles one.";
const POINTER_HANDLER_PREFIXES: [&str; 5] =
    ["onclick", "onpointer", "onmouse", "ontouch", "ondrag"];

#[derive(Debug, Default, Clone)]
pub struct NoInertPointerAffordance;

declare_oxc_lint!(
    /// Disallow pointer cursors on inert native elements.
    NoInertPointerAffordance,
    react_doctor_native,
    correctness,
    version = "0.1.0",
    short_description = "Disallow pointer cursors without interaction.",
);

impl Rule for NoInertPointerAffordance {
    fn should_run(&self, ctx: &ContextHost) -> bool {
        !is_non_production_file(ctx)
    }

    fn run<'a>(&self, node: &AstNode<'a>, ctx: &LintContext<'a>) {
        let AstKind::JSXOpeningElement(opening_element) = node.kind() else {
            return;
        };
        let element_type = resolve_jsx_element_type_name(opening_element, ctx);
        if !HTML_TAG.contains(element_type.as_ref())
            || is_interactive_element(&element_type, opening_element)
        {
            return;
        }
        let Some(class_name) = get_static_class_name(opening_element) else {
            return;
        };
        let tokens = tailwind_class_name_tokens(class_name);
        if get_effective_tailwind_class_name_token(&tokens, |utility| {
            utility.starts_with("cursor-")
        }) != Some("cursor-pointer")
            || element_type == "label"
            || has_pointer_behavior_signal(opening_element)
            || has_wrapping_interaction_boundary(node, ctx)
            || has_nested_interaction_boundary(node, opening_element, ctx)
        {
            return;
        }
        ctx.diagnostic(OxcDiagnostic::warn(MESSAGE).with_label(opening_element.span));
    }
}

fn has_pointer_behavior_signal(opening_element: &oxc_ast::ast::JSXOpeningElement<'_>) -> bool {
    if opening_element
        .attributes
        .iter()
        .any(|attribute| matches!(attribute, JSXAttributeItem::SpreadAttribute(_)))
        || ["tabindex", "ref", "draggable", "contenteditable"]
            .iter()
            .any(|name| has_jsx_prop_ignore_case(opening_element, name).is_some())
    {
        return true;
    }
    if opening_element.attributes.iter().any(|attribute| {
        let JSXAttributeItem::Attribute(attribute) = attribute else {
            return false;
        };
        let oxc_ast::ast::JSXAttributeName::Identifier(identifier) = &attribute.name else {
            return false;
        };
        let attribute_name = identifier.name.as_str();
        POINTER_HANDLER_PREFIXES.iter().any(|prefix| {
            attribute_name
                .get(..prefix.len())
                .is_some_and(|candidate| candidate.eq_ignore_ascii_case(prefix))
        })
    }) {
        return true;
    }
    let Some(JSXAttributeItem::Attribute(role_attribute)) =
        has_jsx_prop_ignore_case(opening_element, "role")
    else {
        return false;
    };
    let Some(role_value) = get_string_literal_attribute_value(role_attribute) else {
        return true;
    };
    role_value
        .trim()
        .split_whitespace()
        .next()
        .is_some_and(|role| is_interactive_role(&role.to_ascii_lowercase()))
}

fn is_interaction_boundary<'a>(
    opening_element: &'a oxc_ast::ast::JSXOpeningElement<'a>,
    ctx: &LintContext<'a>,
) -> bool {
    let element_type = resolve_jsx_element_type_name(opening_element, ctx);
    !HTML_TAG.contains(element_type.as_ref())
        || element_type == "label"
        || is_interactive_element(&element_type, opening_element)
        || has_pointer_behavior_signal(opening_element)
}

fn has_wrapping_interaction_boundary(node: &AstNode<'_>, ctx: &LintContext<'_>) -> bool {
    ctx.nodes()
        .ancestors(node.id())
        .skip(1)
        .any(|ancestor| match ancestor.kind() {
            AstKind::JSXAttribute(_) => true,
            AstKind::JSXElement(element) => is_interaction_boundary(&element.opening_element, ctx),
            _ => false,
        })
}

fn has_nested_interaction_boundary<'a>(
    node: &AstNode<'a>,
    opening_element: &oxc_ast::ast::JSXOpeningElement<'a>,
    ctx: &LintContext<'a>,
) -> bool {
    let parent = ctx.nodes().parent_node(node.id());
    let AstKind::JSXElement(element) = parent.kind() else {
        return false;
    };
    ctx.nodes().iter().any(|candidate| {
        let AstKind::JSXOpeningElement(descendant) = candidate.kind() else {
            return false;
        };
        descendant.span.start >= opening_element.span.end
            && descendant.span.end <= element.span.end
            && is_interaction_boundary(descendant, ctx)
    })
}
